torch/csrc/distributed/c10d/Utils.cu

```diff
@@ -0,0 +1,34 @@
+#include <torch/csrc/distributed/c10d/Utils.hpp>
+
+#include <ATen/Dispatch.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+
+namespace c10d {
+
+__global__ void checkForNaN(float* data, int size) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx < size) {
+        float val = data[idx];
+        // Use device-side assert to check for NaN
+        // This will cause the kernel to terminate if NaN is found
+        CUDA_KERNEL_ASSERT(!isnan(val));
+    }
+}
+
+void checkForNan(const at::Tensor& tensor) {
+  const int blockSize = 256;
+  const int numBlocks = (tensor.numel() + blockSize - 1) / blockSize;
+
+  checkForNaN<<<numBlocks, blockSize>>>(tensor.data_ptr<float>(), tensor.numel());
+  // Wait for the kernel to finish
+  cudaDeviceSynchronize();
```

Comment:
The synchronize may be too expensive- in my mind it's close to the cost of a cpu copy.

Can you explore the options without sync? 

Maybe it's still possible to use this with sync, in cases where nan is repeatedly observed and we can afford to slow down training to find it. But unlikely this way could be enabled during normal training.

Reply:
I don't think we need to sync here as long as we launch the kernel, the DSA/cuda error would block following kernel execution throw run time error

Reply:
Hmm.
I think i believe the DSA would stop future kernels, but it also means the error won't be raised from this call site. Would it be delayed until the next synchronize or caught by the watchdog? That might be fine as long as the error message is understandable and the delivery is reliable.

It would be good to figure out the overhead of the nan check per collective with and without sync, and comment a bit more on the reliability of error delivery. Then we can decide either ship it in one form or the other (or worst case offer a flag for sync/no sync so user can experiment)

Reply:
Error would be araised from the call site like this:
/home/sqzhang/pytorch/torch/csrc/distributed/c10d/Utils.cu:16: checkForNaN: block: [0,0,0], thread: [70,0,0] Assertion `!isnan(data[i])` failed.